File: 01 - Regression/S4 - Polynomial Regression/polynomial_regression.r
# Regresion polinomica

# Import dataset

dataset = read.csv('S4 - Polynomial Regression/Position_Salaries.csv')
dataset = dataset[, 2:3]

# Dividir los datos en conjunto de entrenamiento y conjunto de test
# 
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 2/3)
# training_set = subset(dataset, split == TRUE)
# testing_set = subset(dataset, split == FALSE)

# Ajustar modelo de Regresion Lineal con el conjunto de datos

lin_reg = lm(formula = Salary ~ ., 
             data = dataset)

# Ajustar modelo de Regresion Polinomica con el mismo conjunto de datos

dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4

poly_reg = lm(formula = Salary ~ .,
              data = dataset)

# Visualizacion del modelo lineal

library(ggplot2)

ggplot() +
  geom_point(aes(x = dataset$Level , y = dataset$Salary), 
             color = "red") +
  geom_line(aes(x = dataset$Level , y = predict(lin_reg, newdata = dataset)),
            color = "blue") +
  ggtitle("Prediccion lineal del sueldo en funcion del nivel del empleado") +
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")

# Visualizacion del modelo Polinomico

x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)

ggplot() +
  geom_point(aes(x = dataset$Level , y = dataset$Salary), 
             color = "red") +
  geom_line(aes(x = x_grid , y = predict(poly_reg, newdata = data.frame(Level = x_grid, 
                                                                        Level2 = x_grid^2,
                                                                        Level3 = x_grid^3,
                                                                        Level4 = x_grid^4))),
            color = "blue") +
  ggtitle("Prediccion lineal del sueldo en funcion del nivel del empleado") +
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")

#Prediccion de nuevos resultados con regresion lineal

y_pred_lin = predict(lin_reg, newdata = data.frame(Level = 6.5))

#Prediccion de nuevos resultados con regresion polinomica

y_pred_poly = predict(poly_reg, newdata = data.frame(Level = 6.6, 
                                                     Level2 = 6.5^2,
                                                     Level3 = 6.5^3,
                                                     Level4 = 6.5^4))

File: 01 - Regression/S4 - Polynomial Regression/polynomial_regression.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun 21 17:48:13 2022

@author: pelka
"""

# Regresion Polinomica

# Importar librerias

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el dataset

dataset = pd.read_csv('Position_Salaries.csv')
x = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

"""
# Dividir el data set en conjunto de entrenamiento y conjunto de testing

from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)
"""
 
# Ajustar la regresion lineal con el dataset
 
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(x, y)

# Ajustar la regresion polinomica con el dataset

from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
x_poly = poly_reg.fit_transform(x)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(x_poly, y)

# Visualizacion de los resultados del Modelo Lineal

plt.scatter(x, y, color = "red")
plt.plot(x, lin_reg.predict(x), color ="blue")
plt.title("Modelo de Regresion Lineal")
plt.xlabel("Posicion del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()

# Visualizacion de los resultados del Modelo Polinomico

x_grid = np.arange(min(x), max(x), 0.1)
x_grid = x_grid.reshape(len(x_grid), 1)
plt.scatter(x, y, color = "red")
plt.plot(x_grid, lin_reg_2.predict(poly_reg.fit_transform(x_grid)), color ="blue")
plt.title("Modelo de Regresion Polinomica")
plt.xlabel("Posicion del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()

# Prediccion de nuestros modelos

lin_reg.predict([[6.5]])
lin_reg_2.predict(poly_reg.fit_transform([[6.5]]))

